File: examples/Fortran/Data_Management_unstructured_declare_example.f90
module utils
    use iso_fortran_env, only : REAL64, INT32
    contains
        subroutine normalize_cols(mat, mat_size)
            real    (kind=REAL64), allocatable, dimension(:,:), intent(inout) :: mat
            integer (kind=INT32 )                             , intent(in)    :: mat_size  
            real    (kind=REAL64)                                             :: norm = 0.0_real64
            integer (kind=INT32 )                                             :: i,j
            real    (kind=REAL64), allocatable, dimension(:) :: norms
            !$acc declare device_resident(norms(:))
            allocate(norms(mat_size))
!! Compute the L1 norm of each column
            !$acc parallel loop present(mat(:,:), norms(:))
            do j = 1, mat_size
                norm = 0
                !$acc loop reduction(+:norm)
                do i = 1, mat_size
                    norm = norm + mat(i,j)
                enddo
                norms(j) = norm
            enddo
!! Divide each element of the columns by the L1 norm
            !$acc parallel loop present(mat(:,:), norms(:))
            do j = 1, mat_size
                do i = 1, mat_size
                    mat(i,j) = mat(i,j)/norms(j)
                enddo
            enddo
        end subroutine normalize_cols
end module utils

program normalize
    use utils
    real    (kind=REAL64), allocatable, dimension(:,:)  :: mat
    real    (kind=REAL64)                               :: mat_sum
    integer (kind=INT32)                                :: mat_size=2000
    integer (kind=INT32)                                :: i, j

    allocate(mat(mat_size, mat_size))
    !$acc enter data create(mat)
    call random_number(mat)
    !$acc update device(mat(:,:))
    call normalize_cols(mat, mat_size)
!! Compute the sum of all elements of the matrix
    !$acc parallel loop present(mat(:,:)) reduction(+:mat_sum)
    do j = 1, mat_size
        do i = 1, mat_size
            mat_sum = mat_sum + mat(i,j)
        enddo
    enddo
    !$acc exit data delete(mat)
    deallocate(mat)
    print *, mat_sum, "=", mat_size, "?"
end program normalize

File: examples/C/Data_Management_unstructured_declare_example.c
#include <stdio.h>
#include <stdlib.h>

void normalize_rows(double* mat, size_t size)
{
    double norms[size];
    #pragma acc declare device_resident(norms)
    double norm;
    // Compute the L1 norm of each row
    #pragma acc parallel loop present(mat[0:size*size])
    for (size_t i=0; i<size; ++i)
    {
        norm = 0.;
        #pragma acc loop reduction(+:norm)
        for (size_t j=0; j<size; ++j)
            norm += mat[i*size+j];
        norms[i] = norm;
    }
    // Divide each row element by the L1 norm 
    #pragma acc parallel loop present(mat[0:size*size])
    for (size_t i=0; i<size; ++i)
        for (size_t j=0; j<size; ++j)
            mat[i*size+j] /= norms[i];
}

int main(void)
{
    size_t size = 2000;
    double* mat = malloc(size*size*sizeof(double));
    double sum = 0.;
    srand((unsigned) 12345900);
    for (size_t i=0; i<size; ++i)
        for (size_t j=0; j<size; ++j)
            mat[i*size+j] = (double)rand() / (double) RAND_MAX;
    #pragma acc enter data copyin(mat[0:size*size])

    normalize_rows(mat, size);

    // Compute the sum of all elements in the matrix
    #pragma acc parallel loop present(mat[0:size*size]) reduction(+:sum)
    for (size_t i=0; i<size; ++i)
        for (size_t j=0; j<size; ++j)
            sum += mat[i*size+j];
    #pragma acc exit data delete(mat[0:size*size])
    free(mat);

    printf("%f == %d?\n", sum, size);

    return 0;
}
